deepchem/models/torch_models/infograph.py

```diff
@@ -6,6 +6,269 @@
+        Number of edge features for each input
+    embedding_dim: int
+        Dimension of the embedding
+    num_gc_layers: int
+        Number of graph convolutional layers
+    prior: bool
+        Whether to use a prior expectation in the loss term
+    gamma: float
+        Weight of the prior expectation in the loss term
+    measure: str
+        The divergence measure to use for the unsupervised loss. Options are 'GAN', 'JSD',
+        'KL', 'RKL', 'X2', 'DV', 'H2', or 'W1'.
+    average_loss: bool
+        Whether to average the loss over the batch
+    """
+
+    def __init__(self,
+                 num_features,
+                 embedding_dim,
+                 num_gc_layers=5,
+                 prior=True,
+                 gamma=.1,
+                 measure='JSD',
+                 average_loss=True,
+                 **kwargs):
+        self.num_features = num_features
+        self.embedding_dim = embedding_dim * num_gc_layers
+        self.num_gc_layers = num_gc_layers
+        self.gamma = gamma
+        self.prior = prior
+        self.measure = measure
+        self.average_loss = average_loss
+        self.components = self.build_components()
+        self.model = self.build_model()
+        super().__init__(self.model, self.components, **kwargs)
+
+    def build_components(self) -> dict:
+        return {
+            'encoder':
+                GINEncoder(self.num_features, self.embedding_dim,
+                           self.num_gc_layers),
+            'local_d':
+                MultilayerPerceptron(self.embedding_dim,
+                                     self.embedding_dim, (self.embedding_dim,),
+                                     skip_connection=True),
+            'global_d':
+                MultilayerPerceptron(self.embedding_dim,
+                                     self.embedding_dim, (self.embedding_dim,),
+                                     skip_connection=True),
+            'prior_d':
+                MultilayerPerceptron(self.embedding_dim,
+                                     1, (self.embedding_dim,),
+                                     activation_fn='sigmoid')
+        }
+
+    def build_model(self) -> nn.Module:
+        return InfoGraph(**self.components)
+
+    def loss_func(self, inputs, labels, weights):
+        y, M = self.components['encoder'](inputs)
+        g_enc = self.components['global_d'](y)
+        l_enc = self.components['local_d'](M)
+        local_global_loss = self._local_global_loss(l_enc, g_enc,
+                                                    inputs.graph_index)
+        if self.prior:
+            prior = torch.rand_like(y)
+            term_a = torch.log(self.components['prior_d'](prior)).mean()
+            term_b = torch.log(1.0 - self.components['prior_d'](y)).mean()
+            PRIOR = -(term_a + term_b) * self.gamma
+        else:
+            PRIOR = 0
+        return local_global_loss + PRIOR
+
+    def _local_global_loss(self, l_enc, g_enc, index):
```

Comment:
This should be a function and not a method I think; it doesn't use self

Reply:
It does. It accesses self.device, self.measure and self.average_loss